Add forward compatibility tests in CI

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64139

This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

tugsbayasgalan committed:
[WIP] add forward compatability tests in CI

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -84,6 +84,29 @@ inline bool Argument::isBackwardCompatibleWith(
+  if (lhs->kwarg_only() && !rhs->kwarg_only()) {
+    return false;
+  }
+  if (!rhs->type()->isSubtypeOfExt(lhs->type(), why_not)) {
```
Should this be reversed?

IIUC, `rhs` stands for older version of schema, `lhs` stands for newer version of schema. When `lhs > rhs`, ie newer version of schema allows a type that is a super type of older version of schema, we would have forward compatibility problem. Let me know what you think.

tugsbayasgalan replied:
Yeah you are right. It should be reversed. Will update

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -169,6 +192,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
+        // since they are only used by internal operators
+        && is_vararg() == old.is_vararg() && is_varret() == old.is_varret() &&
+        returns().size() == old.returns().size() &&
+        arguments().size() >= old.arguments().size())) {
```
The case of `arguments.size() < old.arguments().size()` can be FC actually. Think about the case where we remove the last argument, which has a default value.

What do you think?

tugsbayasgalan replied:
That is true. I will update it

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -169,6 +192,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
+    return false;
+  }
+
+  // find the start of out args in the old schema
+  int old_out_start_idx;
+  for (old_out_start_idx = old.arguments().size() - 1; old_out_start_idx > -1;
+       old_out_start_idx--) {
+    if (!old.arguments().at(old_out_start_idx).is_out()) {
+      break;
+    }
+  }
```
Really need to make this into a helper func : )

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -169,6 +192,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
+    if (default_val.isList() || default_val.isGenericDict()) {
+      if (why_not) {
+        *why_not
+            << "Function schema not forward compatible since the new argument '"
+            << arguments().at(i).name() << "' of type "
+            << arguments().at(i).type()->str() << " is a container type.";
```
Add more information here about we don't support default values of container type.

tugsbayasgalan replied:
Sounds good.

gmagogsfm commented on test/test_function_schema.py:
```diff
@@ -112,6 +112,32 @@ def test_string_optional_parameter_default_value(self):
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int c=1) -> Tensor')
+        self.assertTrue(new_schema.is_forward_compatible_with(old_schema))
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int[2] c=1) -> Tensor')
+        self.assertFalse(new_schema.is_forward_compatible_with(old_schema))
```
Looks like `isForwardCompatibleWith`'s PyBind definition doesn't allow returning a string of `why_not`. Could we add it and verify match failures are due to the right reasons?

tugsbayasgalan replied:
That's a good idea. Will do

janeyx99 reviewed (commented):
Looks fine from the CI side!

tugsbayasgalan committed:
Update on "[WIP] add forward compatability tests in CI"


This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

mruberry commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -199,6 +208,28 @@ struct FunctionSchema {
       const FunctionSchema& old,
       std::ostream* why_not = nullptr) const;
 
+  // Checks whether this schema is forward compatible with the old one.
```
Cool list and super clear. Might be helpful to actually state which case is allowed/supported, too, which is adding a new kwarg with a default value at the end of a signature without out= or just before the out= kwarg on signatures that have it

tugsbayasgalan replied:
done

mruberry commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -169,6 +191,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
+            << "Function schema is not forward compatible since the new argument '"
+            << arguments().at(i).name() << "' of type "
+            << arguments().at(i).type()->str() << " has a container type "
+            << "as its' default value.";
```
"its'" -> "its"

tugsbayasgalan replied:
hhha will do :)

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -197,6 +199,49 @@ def check_bc(existing_schemas):
+    new_schema_dict = load_schemas_to_dict()
+    is_fc = True
+    broken_ops = []
+    # TODO skip allow list for now.
```
Can support for the skip list be added to this PR, or are there design issues? Like are you thinking of having a separate BC and FC allow lists?

tugsbayasgalan replied:
The skip list was for BC before, so I was hesitant to include them here. But now I think about it, it probably makes more sense for those schemas to be considered for both back and forward compatibility. I will allow skip list here then. I think we should maintain one single list for both BC and FC. @gmagogsfm What do you think?

mruberry commented on test/test_function_schema.py:
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
         self.assertEquals(schema_a, schema_b)
 
+    def test_forward_compatible_arguments_without_out(self):
+        old_schema = parse_schema('any(Tensor self, int a, int b=1) -> Tensor')
```
Add comments for each case being tested before it's tested

tugsbayasgalan replied:
done

mruberry commented on test/test_function_schema.py:
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int[2] c=1) -> Tensor')
+        is_fc, reason = new_schema.check_forward_compatible_with(old_schema)
+        self.assertFalse(is_fc)
+        self.assertEqual(reason, "Function schema is not forward compatible since the new argument"
```
Is there a good test for type narrowing/widening?

tugsbayasgalan replied:
I added one test where we widen the types of start and end of aten::slice which actually caused production issue before :)

mruberry commented on test/test_function_schema.py:
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
+        new_schema = parse_schema('any(Tensor self, int a) -> Tensor')
+        is_fc, _ = new_schema.check_forward_compatible_with(old_schema)
+        self.assertTrue(is_fc)
+        new_schema = parse_schema('any(Tensor self, int a, int b=1, int c=1) -> Tensor')
```
What about adding a new kwarg without a default value?

tugsbayasgalan replied:
will add a test for it

mruberry commented on test/test_function_schema.py:
```diff
@@ -112,6 +112,53 @@ def test_string_optional_parameter_default_value(self):
+        is_fc, reason = new_schema.check_forward_compatible_with(old_schema)
+        self.assertFalse(is_fc)
+        self.assertEqual(reason, "\'c\' is not forward compatible with the older version of the schema")
+        new_schema = parse_schema('any(Tensor self, int a, int c=1, int b=1) -> Tensor')
```
What about a positional arg becoming kwarg-only and a kwarg-only arg becoming positional?

tugsbayasgalan replied:
good test case. I will add it.

mruberry reviewed (commented):
Overall looks pretty good! I left some comments inline for your review, @tugsbayasgalan. Looking forward to hearing your thoughts!

tugsbayasgalan committed:
Update on "Add forward compatability tests in CI"


This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

tugsbayasgalan committed:
Update on "Add forward compatability tests in CI"


This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

tugsbayasgalan committed:
Update on "Add forward compatability tests in CI"


This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

mruberry reviewed (approved):
LGTM!

tugsbayasgalan committed:
Update on "Add forward compatability tests in CI"


This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

tugsbayasgalan committed:
Update on "Add forward compatibility tests in CI"


This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).

tugsbayasgalan committed:
Update on "Add forward compatibility tests in CI"


This PR adds CI test that can detect forward compatibility breaking changes for aten operator schemas. Since we don't have a strong policy on what constitutes as FC breaking change, we will just warn users when they introduce FC incompatible changes. This will later be replaced with more strict testing once we implement operator versioning. 

Differential Revision: [D30626912](https://our.internmc.facebook.com/intern/diff/D30626912)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30626912).
